src/main/kotlin/org/rust/lang/core/macros/RangeMap.kt

```diff
@@ -0,0 +1,177 @@
+            offset >= s.startOffset && offset < s.endOffset
+        }?.let { (d, s) ->
+            d.startOffset + (offset - s.startOffset)
+        }
+    }
+
+    fun mapOffsetFromCallBodyToExpansion(offset: Int): List<Int> {
+        return ranges.filter { (s, _) ->
+            offset >= s.startOffset && offset < s.endOffset
+        }.map { (s, d) ->
+            d.startOffset + (offset - s.startOffset)
+        }
+    }
+
+    companion object {
+        val EMPTY: RangeMap = CompactRangeMap.EMPTY.unpack()
+
+        fun fromRanges(ranges: List<MappedTextRange>): RangeMap {
+            val textRanges = ranges.map {
+                TextRange(it.srcOffset, it.srcOffset + it.length) to
+                    TextRange(it.offsetInExpansion, it.offsetInExpansion + it.length)
+            }
+            return RangeMap(textRanges)
+        }
+    }
+}
+
+/**
+ * Compact representation of [RangeMap] used for serialization and to save memory.
+ *
+ * Must provide [equals] method because it's used to track changes in macro expansion mechanism
+ */
+@Suppress("DataClassPrivateConstructor")
+data class CompactRangeMap private constructor(private val ranges: List<MappedTextRange>) {
+
+    fun unpack(): RangeMap =
+        RangeMap.fromRanges(ranges)
+
+    fun writeTo(data: DataOutputStream) {
+        data.writeInt(ranges.size)
+        ranges.forEach {
+            data.writeMappedTextRange(it)
+        }
+    }
+
+    companion object {
+        val EMPTY: CompactRangeMap = CompactRangeMap(emptyList())
+
+        fun readFrom(data: DataInputStream): CompactRangeMap {
+            val size = data.readInt()
+            val ranges = (0 until size).map { data.readMappedTextRange() }
+            return CompactRangeMap(ranges)
+        }
+
+        fun from(ranges: SmartList<MappedTextRange>): CompactRangeMap {
+            return CompactRangeMap(ranges.optimizeList())
+        }
+    }
+}
+
+private fun DataInputStream.readMappedTextRange(): MappedTextRange = MappedTextRange(
+    readVarInt(),
+    readVarInt(),
+    readVarInt()
+)
+
+private fun DataOutputStream.writeMappedTextRange(range: MappedTextRange) {
+    writeVarInt(range.srcOffset)
+    writeVarInt(range.offsetInExpansion)
+    writeVarInt(range.length)
+}
+
+data class MappedTextRange(
+    val srcOffset: Int,
+    val offsetInExpansion: Int,
+    val length: Int
+) {
+    init {
+        check(srcOffset >= 0) { "srcOffset should be >= 0; got: $srcOffset" }
+        check(offsetInExpansion >= 0) { "offsetInExpansion should be >= 0; got: $offsetInExpansion" }
+        check(length > 0) { "length should be grater than 0; got: $length" }
+    }
+}
+
+/**
+ * If receiver element is inside a macro expansion, returns the leaf element inside the macro call
+ * from which the first token of this element is expanded. Always returns an element inside a root
+ * macro call, i.e. outside of any expansion.
+ */
+fun PsiElement.findElementExpandedFrom(): PsiElement? {
```

Comment:
I'd prefer to move all these helper functions and extension methods to `RsExpandedElement.kt` to group all related methods in one place